manuals/nk2export.1
.Dd June 25, 2026
.Dt NK2EXPORT 1
.Os
.Sh NAME
.Nm nk2export
.Nd exports items stored in a Nickfile (NK2)
.Sh SYNOPSIS
.Nm nk2export
.Op Fl c Ar codepage
.Op Fl l Ar log_file
.Op Fl t Ar target
.Op Fl hqvV
.Ar source
.Sh DESCRIPTION
.Nm nk2export
is a utility to export items stored in a Nickfile (NK2)
.Pp
.Nm nk2export
is part of the
.Nm libnk2
package.
.Nm libnk2
is a library to access the Nickfile (NK2) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl l Ar log_file
specify the file in which to log information about the exported items
.It Fl q
quiet shows minimal status information
.It Fl t Ar target
specify the target directory to export to (default is the source filename \
followed by .export)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# nk2export Outlook.NK2
nk2export 20260625
.sp
Opening file.
Exporting aliases.
Exporting alias 1 out of 1.
Export completed.
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr nk2info 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libnk2/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* nk2export check if always adding .export is the desired behavior
* libnk2_io_handle.c check if LIBNK2_VALUE_TYPE_BOOLEAN is working as intended
* item values
  - implement currency as: LIBFVALUE_VALUE_TYPE_FIXED_POINT_64BIT
    what about location of the point ?
* in io_handle_read_items used set array value instead of append ?

Optimizations:
* map (file offset and size) items and item values first read on demand ?
* pre allocate item array entries

library
* additional value type support functions
  - filetime
* what about multiple alias entries 0x6001 of a single item?
  - they are set in values table but identifier lookup will return first only
* local library support

Tools
* add signal handling
* nk2export
  - add export mode like pffexport
  - export alias files instead of item values
  - export to CSV/TSV, HTML, XML
  - add -d option to dump item values and use this in test_nk2export.sh
* nk2info

Beta release
* remove deprecated definition in include/libnk2/mapi.h

20090128
* Created initial version based on libpff for the library structure and common code
* see `git log' for more recent change log



nk2tools/nk2export.c
/*
 * Extracts items from a Nickfile (NK2)
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <narrow_string.h>
#include <system_string.h>
#include <types.h>
#include <wide_string.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "log_handle.h"
#include "nk2input.h"
#include "nk2tools_getopt.h"
#include "nk2tools_libcerror.h"
#include "nk2tools_libclocale.h"
#include "nk2tools_libcnotify.h"
#include "nk2tools_libcpath.h"
#include "nk2tools_libnk2.h"
#include "nk2tools_output.h"
#include "nk2tools_signal.h"
#include "nk2tools_unused.h"

export_handle_t *nk2export_export_handle = NULL;
int nk2export_abort                      = 0;

/* Signal handler for nk2export
 */
void nk2export_signal_handler(
      nk2tools_signal_t signal NK2TOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function   = "nk2export_signal_handler";

	NK2TOOLS_UNREFERENCED_PARAMETER( signal )

	nk2export_abort = 1;

	if( nk2export_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     nk2export_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use nk2export to export items stored in a Nickfile (NK2).";

	nk2tools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'l', "log_file", "logs information about the exported items" },
		{ 'q', NULL, "quiet shows minimal status information" },
		{ 't', "target", "specify the target directory to export to (default is the source filename followed by .export)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error               = NULL;
	log_handle_t *log_handle               = NULL;
	system_character_t *log_filename       = NULL;
	system_character_t *option_codepage    = NULL;
	system_character_t *option_target_path = NULL;
	system_character_t *path_separator     = NULL;
	system_character_t *source             = NULL;
	char *program                          = "nk2export";
	system_integer_t option                = 0;
	size_t source_length                   = 0;
	uint8_t print_status_information       = 1;
	int number_of_options                  = (int) ( sizeof( options ) / sizeof( nk2tools_option_t ) );
	int result                             = 0;
	int verbose                            = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "nk2tools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( nk2tools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	nk2tools_output_version_fprint(
	 stdout,
	 program );

	if( nk2tools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = nk2tools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				nk2tools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				nk2tools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'q':
				print_status_information = 0;

				break;

			case (system_integer_t) 't':
				option_target_path = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				nk2tools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		nk2tools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	if( option_target_path == NULL )
	{
		source_length = system_string_length(
		                 source );

		path_separator = system_string_search_character_reverse(
		                  source,
		                  (system_character_t) LIBCPATH_SEPARATOR,
		                  source_length );

		if( path_separator == NULL )
		{
			path_separator = source;
		}
		else
		{
			path_separator++;
		}
		option_target_path = path_separator;
	}
	libcnotify_verbose_set(
	 verbose );
	libnk2_notify_set_stream(
	 stderr,
	 NULL );
	libnk2_notify_set_verbose(
	 verbose );

	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( export_handle_initialize(
	     &nk2export_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( nk2export_export_handle != NULL );
#endif
	nk2export_export_handle->print_status_information = print_status_information;

	if( option_codepage != NULL )
	{
		result = export_handle_set_ascii_codepage(
		          nk2export_export_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in export handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( export_handle_set_target_path(
	     nk2export_export_handle,
	     option_target_path,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set target path.\n" );

		goto on_error;
	}
	result = export_handle_create_items_export_path(
	          nk2export_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create items export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 nk2export_export_handle->items_export_path );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	fprintf(
	 stdout,
	 "Opening file.\n" );

	if( export_handle_open(
	     nk2export_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	result = export_handle_export_file(
	          nk2export_export_handle,
	          log_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file.\n" );

		goto on_error;
	}
	if( export_handle_close(
	     nk2export_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close export handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &nk2export_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log handle.\n" );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	if( result == 0 )
	{
		fprintf(
		 stdout,
		 "No aliases to export.\n" );
	}
	else
	{
		fprintf(
		 stdout,
		 "Export completed.\n" );
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( nk2export_export_handle != NULL )
	{
		export_handle_free(
		 &nk2export_export_handle,
		 NULL );
	}
	if( log_handle != NULL )
	{
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

